boot: add submenu support

TobiPeterG opened the pull request:
This also shows the current menu path
~Fixes a bug displaying status messages~
-------------------------------------------------------------------
I tried multiple implementation methods, i like this one the most. While it introduces a new struct and, union and enum, it allows easier expandability in the future. Also, since parent entries are also saved, this approach with saving the type of entry makes entry handling much easier, as one only has to check for the type instead of calculating with indices.

I added some helper functions that  might not fit in this file (e.g. find_last_substring, efi_wcslen, efi_wcschr). Should I add them to util.c instead?

Menu titles can be set either by the directory name (default) or by defining a .menu file in the directory, it currently only supportes the title key, but this allows setting a custom name.
As this approach allows nestes submenu structures, I thought it might be useful to show the title and reletive path of the current menu. I think this significantly increases the UX. Also, the path is trimmed if it doesn't fit on the screen to remove previous menu titles and not lose the current one.

I also tried to make existing code more readable by either simplifying it or exporting functionality into extra functions (especially for code repititions).

I tried to handle memory allocations correctly, if not, please tell me so that I can fix it, I don't want to cause memory leaks.

I used recursive call where possible to prevent code duplication.

I never contributed to systemd, I tried to follow the code style guidelines and the code style of this file. Should you not like a part, I'll try to re-work it.

~I also fixed an existing bug that prevented the last menu entry from being shown when there are more entries than the display can show, I fixed that as well (the y_start variable was used wrongly).~ (There is no bug)

While I tried to test everything, I can't guarantee that I catched every edge case. As this also introduces quite a bit of code, extensive testing would be helpful to prevent unbootable systems later on.
(Though I really tried everything to make it break, but I also trusted the existing implementation to be robust).

This feature is useful for e.g. OpenSUSE, as it supports Snapshots, but adding all of them to the main menu would only clutter it (there's already work on implementing snapshot support).
Other tools like grub-btrfs might also be adapted for systemd-boot with submenus.

I added some comments but not too many, as code comments itself at best. Should any code be hard to understand, I can try to rework it.

As I don't know if this feature is even wanted at all / in its current form, I marked this as draft for now.

Are there automated tests I can execute / should add or is there a linter I can execute?

Should close https://github.com/systemd/systemd/issues/9095

medhefgo commented on src/boot/efi/boot.c:
```diff
@@ -612,105 +677,222 @@ static EFI_STATUS reboot_into_firmware(void) {
+        line[j] = title_show[k];
+    for (; j < line_width; j++)
+        line[j] = ' ';
+    line[line_width] = '\0';
```
While you're touching all this: This kind of thing should be doable with something like `xasprintf(%*.*s);` etc. I'd prefer going away from this kind of string manipulations.

TobiPeterG replied:
Should this happen here or in another PR?

medhefgo replied:
Whatever you think works best for you. It's okay for this PR to contain refactors, just don't squash it into a single commit.

TobiPeterG replied:
Alright, I'll split it into multiple commits 👍🏻

TobiPeterG replied:
see https://github.com/systemd/systemd/pull/31529

medhefgo commented on src/boot/efi/boot.c:
```diff
@@ -2563,6 +3023,72 @@ static void export_variables(
+}
+
+/* EFI quicksort implementation */
+void efi_qsort(Entry** array, size_t low, size_t high, int (*cmp)(const Entry*, const Entry*)) {
```
We have sort_pointer_array for that.

TobiPeterG replied:
I noticed that the sort_pointer_array function uses bubble sort, which is fine, but not the fastest. That's why I thought using quick sort might have some benefits.
Would it be alright to replace bubble sort with quick sort? It allowed me to specify the range in which the entries should be sorted, but I didn't know if sort_pointer_array was used by another function as well, that's why I didn't change it directly.

medhefgo replied:
Let's not. Look again, it's insertion sort and that's enough. Unless you can find me a valid use-case for hundreds (probably more like thousands) of boot entries where sorting will become a bottle-neck.

TobiPeterG replied:
Ah lol you are right, I trusted ChatGPT there
Then I'll not change it 👍🏻

medhefgo reviewed (commented):
> I also fixed an existing bug that prevented the last menu entry from being shown when there are more entries than the display can show, I fixed that as well (the y_start variable was used wrongly).

That sounds weird. I made sure this all works correctly the last time I touched it...

- Please add/split out helper functions in separate commits (and anything else that's a logically contained change too).
- Any (new) string-helper functions should go into `efi-string.c` and get a userspace test in `test-efi-string.c`.
- Use `char16_t` und `u""` only for efi strings.
- Just use `clang-format` on any lines you touch (instead of whole file) to fix the coding style. (`ctr+k, ctrl+f` in vscode, your editor may vary.)

poettering commented on src/boot/efi/boot.c:
```diff
@@ -43,7 +43,26 @@ enum loader_type {
 typedef struct {
+    EntryType entry_type;
+    EntryData entry_data;
+} Entry;
```
coding style. use 8space identation. .editorconfig, the emacs + vim files we have in the tree should get that right btw.

poettering commented on src/boot/efi/boot.c:
```diff
@@ -1098,10 +1195,82 @@ static void config_entry_free(ConfigEntry *entry) {
+                config_entry_free(entry->entry_data.config_entry);
+        } else if (entry->entry_type == MENU) {
+                menu_free(entry->entry_data.menu);
+        }
```
coding style. no {} around single line if blocks

poettering commented on src/boot/efi/boot.c:
```diff
@@ -1098,10 +1195,82 @@ static void config_entry_free(ConfigEntry *entry) {
+                menu_free(entry->entry_data.menu);
+        }
+
+        // Finally, free the entry itself
```
coding style. we use /* */ for explanatory comments, and // for fixme comments only.

keszybz commented on src/boot/efi/boot.c:
```diff
@@ -778,16 +778,16 @@ static bool menu_run(
                         size_t x = (x_max - len) / 2;
                         status[len] = '\0';
-                        print_at(0, y_status, COLOR_NORMAL, clearline + x_max - x);
+                        print_at(0, y_status + 1, COLOR_NORMAL, clearline + x_max - x);
```
Please add an explanation in the commit message what is being fixed. E.g., maybe an example or some other explanation.

I would like to test this patch, but since I'm not sure what the bug is, and how it's supposed to be fixed, that is hard.

TobiPeterG replied:
Hey,
thank you very much for your comment :)
The bug I encountered was that the status line was not correctly shown. It was off by one line. But it might be that I introduced this issue myself?

keszybz commented on src/boot/efi/boot.c:
```diff
@@ -618,6 +622,53 @@ static EFI_STATUS reboot_into_firmware(void) {
+
+        if (entry->group_entry == NULL)
+                return true;
+        else if (entry->group_entry == visible_group)
```
Please drop the `else` here… It doesn't matter much, but that's the usual coding style.

TobiPeterG replied:
done

keszybz commented on src/boot/efi/boot.c:
```diff
@@ -665,58 +718,64 @@ static bool menu_run(
+                        for (size_t i = 0; i < config->n_entries; i++){
+                                if(entry_is_visible(config->entries[i], visible_group))
+                                        line_width = MAX(line_width, strlen16(config->entries[i]->title_show));
+                        }
```
Coding style: no {} when not needed, as space between `if` and `(`.

TobiPeterG replied:
This seems to be outdated now :)
I tried to check everything, but it might be that I missed parts

keszybz commented on src/boot/efi/boot.c:
```diff
@@ -665,58 +718,64 @@ static bool menu_run(
-
+                        lines = xnew(char16_t *, visible_entries + 1);
+                        size_t idx_lines = 0;
                         for (size_t i = 0; i < config->n_entries; i++) {
```
Just say:
```C
for (size_t i = 0, idx_lines = 0; …
```

TobiPeterG replied:
done

keszybz reviewed (changes requested):
I started reviewing the code… But I don't think we're at the point where this makes sense yet.

Please first describe what the design is and how this is supposed to be used. This really must be part of the commit message.

In general, the code is vastly overengineered. Once we have figured out the actual design, I'm sure we can come up with a much simpler implementation.

poettering commented on src/boot/efi/boot.c:
```diff
@@ -618,6 +622,53 @@ static EFI_STATUS reboot_into_firmware(void) {
+static bool entry_is_visible(ConfigEntry *entry, ConfigEntry *visible_group) {
+        assert(entry);
+
+        if (entry->group_entry == NULL)
```
as per coding style, when we check for non-NULL or NULL we don't do explicit `== NULL` and `!= NULL` checks, but simply check relying on C's downgrade to bool feature.

TobiPeterG replied:
done

poettering commented on src/boot/efi/boot.c:
```diff
@@ -618,6 +622,53 @@ static EFI_STATUS reboot_into_firmware(void) {
+        return idx;
+}
+
+static size_t entries_count(Config *config, ConfigEntry *visible_group) {
```
maybe call this entries_count_visible()

TobiPeterG replied:
done

poettering commented on src/boot/efi/boot.c:
```diff
@@ -1878,6 +1964,85 @@ static void config_title_generate(Config *config) {
+        for (; i < group_count; i++) {
+                _cleanup_(config_entry_freep) ConfigEntry *group_entry = NULL;
+                group_entry = xnew(ConfigEntry, 1);
+                *group_entry = (ConfigEntry){ .title = xstrdup16(u"More..."), .group_entry = NULL, .type = LOADER_MORE, .id = xstrdup16(u"More...") };
```
please add line breaks.

TobiPeterG replied:
Is there a linter / tool to check which lines would need a line break?

lnussel replied:
If your editor reads `.editorconfig` it would limit the line length automatically. The file at hand violates it at several places already though.

poettering commented on src/boot/efi/boot.c:
```diff
@@ -1849,6 +1935,98 @@ static void generate_boot_entry_titles(Config *config) {
+                if(!config->entries[i]->sort_key)
+                        continue;
+                size_t k = i;
+                for(; k < config->n_entries; k++){
```
coding style, please add space between `for` and `(`.

poettering commented on src/boot/efi/boot.c:
```diff
@@ -1849,6 +1935,98 @@ static void generate_boot_entry_titles(Config *config) {
+                                continue;
+                        break;
+                }
+                if(k>=i+2){
```
same here. also spaces between `k` and `>=` and `i+2`

poettering commented on src/boot/efi/boot.c:
```diff
@@ -1849,6 +1935,98 @@ static void generate_boot_entry_titles(Config *config) {
+                                break;
+                        if(!config->entries[k]->sort_key)
+                                break;
+                        if(strcmp(config->entries[i]->sort_key,
```
similar here.

poettering replied:
(and everywhere else)

poettering commented on src/boot/efi/boot.c:
```diff
@@ -1849,6 +1935,98 @@ static void generate_boot_entry_titles(Config *config) {
+                        if(!config->entries[k]->sort_key)
+                                break;
+                        if(strcmp(config->entries[i]->sort_key,
+                                  config->entries[k]->sort_key) == 0)
```
please use streq()

poettering commented on src/boot/efi/boot.c:
```diff
@@ -1849,6 +1935,98 @@ static void generate_boot_entry_titles(Config *config) {
+        if(group_count == 0)
+                return;
+        config_create_groups(config, indices, group_count);
+        free(indices);
```
please use `_cleanup_free_` instead of manual freeing

poettering reviewed (changes requested):
i did some superficial reviews, but please rebase. also, please squash your commits. we want "perfect" PRs, i.e. where each commit is a logical step, but not necessarily a historical one. please do not do commits that fix up mistakes in prior commits in the very same PR.

keszybz commented on src/boot/boot.c:
```diff
@@ -1791,6 +1796,86 @@ static void generate_boot_entry_titles(Config *config) {
+typedef struct {
+        size_t start_index;
+        size_t end_index;
+} Indices;
```
This is really strange.

TobiPeterG replied:
Why?
I need a way to track the start and end indices of entry groups (see below). The start and end index of a group always belong together and are "worthless" without the other, so my OOP patterns kicked in to create the struct to also always save the start and end index of a group in the same structure. :)
Is this not "the C way"?

keszybz replied:
Yeah. It's not needed. Just put the two indices directly in the embedding struct.

keszybz commented on src/boot/boot.c:
```diff
@@ -1791,6 +1796,86 @@ static void generate_boot_entry_titles(Config *config) {
+} Indices;
+
+static void config_copy_entries(
+                Config *config, BootEntry **entries, size_t start_index, size_t end_index, bool has_group_entry) {
```
This is very hard to understand.

TobiPeterG replied:
I admit that this function is hard to understand without the context of the other commits and how the code works. I will try to explain the code, I hope the comment is helpful when looking at the code:

I guess it's easier to start from bottom to top.
The idea is: In the first step in the config_group_entries function, we detect the groups we have (so how many entries have the same sort key). This is thankfully quite easy, as

a) the entries are already sorted by sort-key when my function is invoked (as described in https://uapi-group.org/specifications/specs/boot_loader_specification/#sorting) and

b) every entry without a sort-key should be at the very bottom. If 2 entries or less exist with the same sort-key, they shouldn't be grouped at all (as this would lead to equal/more entries in the menu, not less).

So, we only save those groups we actually want to group (those with more than 2 entries). Of these groups, we ignore the first entry, as we won't copy it to the submenu (the newest entry should be always shown, which is, conveniently, always the first one, as the list is sorted). See below for my reasons for this design decision.

Now we know the groups.
The next step is to actually create the group now that we know which groups exist, for this we have the config_create_groups function. It uses the helper function config_copy_entries to create a new list of entries. So, we save the old one and create an empty one that we will fill. It might be that the first entries in the boot menu do not form a group that we want to collapse, so we can just copy all entries until the first group starts without checking.
This will already include the first entry of the first actual group (as we don't want the first entry to be collapsed/hidden). When the first entries start that we want to copy, we assign the group entry. But since this commit only introduces the logic to detect groups, we always set the group entry (so the "More..." entry from the demo video below) to NULL and no group entry is created at all. So all the config_copy_entries function does with this commit is... copying the old entries and not changing them in any way, to not create visual/functional changes. The has_group_entry parameter is a leftover that I need to remove from this commit, it will later be used to determine if a group entry should be created for entries or not (because for all entries that should not be collapsed, it will be unset, for all entries that should be collapsed, it will be set).

After we copied all entries before the first group, there will be at least one group to be collapsed. So, we copy their entries as well (and later set their group entry). Now, we are either at the end of the list, entries without groups follow or another group follows. If entries that do not belong to a group follow, we just copy them (later with the has_group_entry bool set to false) and now, again, we are either at the end of the list or a group follows. We copy them (with has_group_entry set to true) and set their group entry etc..

And that's actually it for this commit. We basically just detect groups and implement the logic that _could_ create the collapsable subgroups.

---------------------------------------------------------

Setting the group indices could also be done by saving all entries with the same sort-key, no matter how many entries belong to the group. But we then need to later always check if the next entry already belongs to a group, if no, copy it, if yes, we only directly copy it when it's the first entry of the group (as we don't want it to be collapsed) and then copy the others, adding  a lot of logic in the config_create_groups function that we don't need now. We know that all entries within indices are to be collapsed, without exception and always have a group entry created above them (introduced in later commits).

Another design choice that could be done differently is introducing the linkage to the group entry. Technically, it's not necessary, as we could, when hitting enter on a group entry in the actual boot menu, have a look at the menu entry above, take its sort-key and check for all entries below the group entry if the sort-key is equal and only show/collapse those. If this is the desired solution, I can, of course, switch to that. I just thought that it introduces a very small performance hit (though of course neglectable), as we need to check an additional entry when showing/collapsing entries and adds additional logic in the function to show the menu. As we already know which entries will be collapsed together anyway in the detection phase introduced in the first commit, we can already use it and save some time later on.
I hope this helps. :)

Should this text be added to the commit message? I guess that would be quite long :D

I could also offer to have a short zoom meeting to discuss the code, so you can directly ask me why I did things the way I did them and you can directly point to parts that need to be improved. :)

I see that the code might not be easy to understand at first glance, but I tried to keep it performant and it will make more sense with the following commits (I hope :P) :)

keszybz replied:
Thank you for this explanation. It really helps to understand the design.

I think this design is more complicated than it has to be. We are dealing with a handful of entries, and it's totally fine to just iterate over them and do any display processing on the fly.

Please consider the following design instead:
- Sometime before displaying the menu, iterate over entries, and figure out if grouping should be done and which entries to show the top level (IIUC, this would either be the first entry or the one marked as default. The code from the last commit already does this.)  The Entry struct can be extended with `group` and `is_group_primary` fields (naming TBD).
- Add a "global" `current_group` field that is similar to the current index field
- When displaying all entries, if grouping is enabled, for each group, show the entries that are marked with `is_group_primary`. Insert a "more <group>" row after the group.
- When asked to display a group, skip entries that don't have the right group marker.

EDIT: I see that @poettering made similar comments above. (I didn't see them previously because they were hidden by github…). His design is a bit different. The details are something that needs to be figured out when implementing this. But the basic idea is the same: have a flat list, display parts of it on the fly as needed.

EDIT2: It might be OK to insert the "more" entries into the entry list, if that makes the implementation easier. Then basically for a specific group, if it is marked as "expanded", you'd show all entries and skip the "more" entry, and if it collapsed, show the one or two primary entries and "more" and skip the rest.

keszybz reviewed (commented):
This patchset cannot be reviewed. Each commit has a single line description that doesn't explain anything.

For a patchet that adds menu support, we really need a preview of the submenus, or a detailed description. If a commit is about visual changes, then those visual changes need to be described.

yuwata reviewed (changes requested):
Please break commit message with suitable width. Currently it is overly long:
![Screenshot_2025-05-25_09-55-16](https://github.com/user-attachments/assets/7520f667-417b-4a03-bb6c-c4dab24fcafb)

Also, as requested by @keszybz, please extend commit message.

keszybz commented on src/boot/boot.c:
```diff
@@ -61,6 +61,7 @@ typedef enum LoaderType {
         LOADER_BAD,           /* Marker: this boot loader spec type #1 entry is invalid */
         LOADER_IGNORE,        /* Marker: this boot loader spec type #1 entry does not match local host */
         _LOADER_TYPE_MAX,
+        LOADER_MORE,
```
This is very strange. `_LOADER_TYPE_MAX` should always be last. And also please add a comment.

keszybz commented on src/boot/boot.c:
```diff
@@ -96,7 +97,9 @@ static const char *reboot_on_error_table[_REBOOT_ON_ERROR_MAX] = {
 DEFINE_PRIVATE_STRING_TABLE_LOOKUP_TO_STRING(reboot_on_error, RebootOnError);
 
-typedef struct BootEntry {
+typedef struct BootEntry BootEntry;  /* Forward declaration */
```
This comment is not needed. Forward declarations are a basic C concept…

keszybz commented on src/boot/boot.c:
```diff
@@ -120,7 +123,8 @@ typedef struct BootEntry {
         char16_t *next_name;
         unsigned profile;
-} BootEntry;
+        BootEntry *group_entry;
```
A comment here though would be nice.

keszybz commented on src/boot/boot.c:
```diff
@@ -1430,6 +1434,7 @@ static void boot_entry_add_type1(
         }
 
         entry->device = device;
+        entry->group_entry = NULL;
```
This is not needed.

keszybz commented on src/boot/boot.c:
```diff
@@ -1874,6 +1879,86 @@ static void generate_boot_entry_titles(Config *config) {
+                config->entries[config->n_entries - 1]->group_entry = group_entry;
+        }
+        TAKE_PTR(group_entry);
+}
```
Any copying of entries seems completely unnecessary.

keszybz reviewed (changes requested):
Sorry for the slow review. 

The current design with copying entries is overcomplicated and it cascades into the need for the helper functions and adjustments. I'm pretty sure this can be done much more simply.

Please collapse all the commits into a single one. They don't make sense separately.

TobiPeterG committed:
boot: add grouped menu entries with expandable rows

Build collapsible menu groups from the final sorted entry list and keep the
grouping metadata local to menu rendering. Visible rows are rebuilt per menu
refresh so number keys, hotkeys, highlighting, and default markers all operate
on what is actually shown on screen.

When multiple entries share a sort key, keep the newest entry visible and also
show the currently default/preferred entry if it belongs to the same
group. Add an expandable "Show More..." / "Show Less..." row to reveal the
remaining entries in that group.

Signed-off-by: TobiPeterG <github.threefold020@passmail.net>
